File: Day 4/Task 1/Task 1.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <string>
#include <utility>


//check if one of the two ranges contains the other
bool FullyInclusiveRanges(const std::pair<int, int> &firstInterval, const std::pair<int, int> &secondInterval) {
	if (secondInterval.first >= firstInterval.first && secondInterval.second <= firstInterval.second)
		return true;

	else if ((firstInterval.first >= secondInterval.first && firstInterval.second <= secondInterval.second))
		return true;

	return false;
}

int main()
{
	std::pair<int, int> firstInterval, secondInterval;
	std::ifstream fhandle("day_4_1.txt");

	std::string line;
	std::regex pattern("([0-9]+)\-([0-9]+),([0-9]+)\-([0-9]+)");
	std::smatch ranges; // match object containing the matched pattern and possible captures

	int numOfFullyInclusive = 0;
	while (std::getline(fhandle, line)) {
		if (std::regex_search(line, ranges, pattern)) {

			firstInterval = std::make_pair(std::stoi(ranges[1]), std::stoi(ranges[2]));
			secondInterval = std::make_pair(std::stoi(ranges[3]), std::stoi(ranges[4]));

			if (FullyInclusiveRanges(firstInterval, secondInterval))
				numOfFullyInclusive++;
			
		}
	}
	std::cout << "Number of assignment pairs where one range fully contains the other: " << numOfFullyInclusive;



}

File: Day 4/Task 1/Task_1.py
import re

def fully_inclusive_ranges(firstInterval, secondInterval):
    if secondInterval[0] >= firstInterval[0] and secondInterval[1] <= firstInterval[1]:
        return True
    elif firstInterval[0] >= secondInterval[0] and firstInterval[1] <= secondInterval[1]:
        return  True
    return False


num_of_fully_inclusive = 0

with open("day_4_1.txt") as fhandle:
    for line in fhandle:
        line = line.rstrip()
        bounds = re.findall("([0-9]+)\-([0-9]+)", line)
        if bounds is not None:
            bounds = [(int(x), int(y)) for x, y in bounds]

            if fully_inclusive_ranges(bounds[0], bounds[1]):
                num_of_fully_inclusive += 1

print('Number of assignment pairs where one range fully contains the other: ', num_of_fully_inclusive)


